File: algo_2021/s6.-dfs-comb-perm/18.Subsets_II/18_subsets_ii__hash_table.java
public class Solution {
    /**
     * @param nums: A set of numbers.
     * @return: A list of lists. All valid subsets.
     */
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>> subsets = new ArrayList<>();
        HashMap<String, Boolean> visited = new HashMap<String, Boolean>();
        Arrays.sort(nums);
        dfs(nums, 0, new ArrayList<>(), subsets, visited);
        
        return subsets;
    }
    
    String getHash(List<Integer> subset) {
        String hashString = "";
        for (int i = 0;i < subset.size(); i++) {
            hashString += subset.get(i).toString();
            hashString += "_";
        }
        return hashString;
    }
    
    void dfs(int[] nums, 
             int startIndex, 
             List<Integer> subset,
             List<List<Integer>> subsets,
             HashMap<String, Boolean> visited) {
        String hashString = getHash(subset);
        
        if (visited.containsKey(hashString)) {
            return ;
        }
        
        visited.put(hashString, true);
        subsets.add(new ArrayList<Integer>(subset));
        for (int i = startIndex;i < nums.length; i++) {
            subset.add(nums[i]);
            dfs(nums, i + 1, subset, subsets, visited);
            subset.remove(subset.size() - 1);
        } 
    } 
}

File: lintcode_algo_2021/s6.-dfs-comb-perm/18.Subsets_II/18_subsets_ii__hash_table.py
class Solution:
    """
    @param nums: A set of numbers.
    @return: A list of lists. All valid subsets.
    """
    def subsetsWithDup(self, nums):
        # write your code here
        res = []
        visited = {}
        nums.sort()
        self.dfs(nums, 0, [], res, visited)
        return res
    def get_hash(self, subset):
        hash_string = ""
        for item in subset:
            hash_string += str(item)
            hash_string += "_"
        return hash_string
        
    def dfs(self, nums, index, subset, res, visited):
        hash_string = self.get_hash(subset)
        if hash_string in visited:
            return 
        visited[hash_string] = True
        res.append(list(subset))
        for i in range(index, len(nums)):
            subset.append(nums[i])
            self.dfs(nums, i + 1, subset, res, visited)
            subset.pop()